modernize test_sparse

aocsa opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54572 modernize test_sparse**

Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

aocsa committed:
modernize test_sparse

[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

Looking forward your comments cc @mruberry


[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

Looking forward your comments cc @mruberry


[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

Looking forward your comments cc @mruberry


[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

Looking forward your comments cc @mruberry


[ghstack-poisoned]

mruberry commented on torch/testing/__init__.py:
```diff
@@ -297,6 +297,9 @@ def __add__(self, other):
 def floating_types():
     return _floating_types
 
+def floating64_type_and(*dtypes):
```
Per the note above on line 282, these functions are intended to correspond 1:1 with the C++ dispatch macros, so let's not add a new function here.

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks 


[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -974,20 +974,20 @@ def setUp(self):
 
         v_size = [nnz] + list(size[sparse_dim:])
-        v = torch.randn(*v_size, device=device)
+        v = torch.randn(*v_size, dtype=dtype, device=device)
```
make_tensor() here might be nice because it will support more types

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -974,20 +974,20 @@ def setUp(self):
         set_rng_seed(SEED)
 
-    def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device='cpu'):
+    def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device='cpu', dtype=torch.double):
```
I'd suggest requiring device and dtype, like make_tensor(), but if that's a pain for this PR then can we make the default 'cpu' and 'torch.float'?

aocsa replied:
Sure, I have a question, why the default `dtype` value should be `torch.float` if the default dtype for test_sparse and test_autograd  is `torch.double`
https://github.com/pytorch/pytorch/blob/0e03a2978ad5ed2be09f3e4691b2d04d6d4c6c90/test/test_sparse.py#L5

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2089,3 +2089,23 @@ def inner(self, *args, **kwargs):
+
+# This is a wrapper that wraps a test to run this test twice, one with
+# coalesced=True, another with coalesced=False for coalesced/uncoalesced sparse tensors.
+def coalescedonoff(f):
```
This approach seems OK but I think the decorator can be simplified and made more future-proof by passing *args **kwargs to the test, and just passing `coalesced` as an additional kwarg-only argument to the test.

So a test could have a signature:

```
test_foo(self, device, dtype, *, coalesced)
```

For example, and the varargs would pass through as expected.

What would you think of that tweak?

mruberry replied:
(This would also mean the checks for nargs could be removed.)

aocsa replied:
Agree, this could be simplified in few lines of code using `*args` and `**kwargs`

mruberry commented on test/test_sparse.py:
```diff
@@ -30,60 +30,28 @@
-        self.index_tensor = lambda *args: torch.tensor(*args, dtype=torch.int64, device=self.device)
-        self.value_empty = lambda *args: torch.empty(*args, dtype=self.value_dtype, device=self.device)
-        self.value_tensor = lambda *args: torch.tensor(*args, dtype=self.value_dtype, device=self.device)
+        self.index_tensor = lambda *args, **kwargs: torch.tensor(*args, **kwargs, dtype=torch.int64)
```
This is an interesting style to assign a function to an attribute of this class.

Why isn't this a helper function?

aocsa replied:
It could be a helper function, previously `self.index_tensor` was just an alias for int64 tensors for specific `self.device` value. Now is the same except that  device value can change.

mruberry commented on test/test_sparse.py:
```diff
@@ -107,9 +75,17 @@ def randn(self, *args, **kwargs):
-        return self.value_empty(*args, **kwargs).normal_()
+        return torch.empty(*args, **kwargs).normal_()
+
+    @dtypes(torch.double)
```
Prefer torch.float if using a single floating point type (it's OK to not change this in this PR if it would be a pain)

aocsa replied:
I was trying to preserve the same test behavior, in previous test_sparse, the value_dtype is torch.float64;
https://github.com/pytorch/pytorch/blob/0e03a2978ad5ed2be09f3e4691b2d04d6d4c6c90/test/test_sparse.py#L60 

However I can change that, I understand that torch.float is prefered because it will make tests lighter

mruberry commented on test/test_sparse.py:
```diff
@@ -292,17 +274,17 @@ def fn(x):
-            v = self.value_tensor([2, 1, 3, 4]).to(dtype=dtype)
-            x = self.sparse_tensor(i, v, torch.Size([3, 4, 5]))
-            res = self.value_tensor([
+        for value_type in [torch.double]:
```
Shouldn't this just set value_type = dtype?

aocsa replied:
Sure, my mistake

mruberry reviewed (approved):
Hey @aocsa! Because this is such a large PR I'd like to land it quickly, so I propose we stop editing it now, get it landed, and then we can deal with the few follow-up issues separately. Does that sound good to you?

It's great to see a more modern version of test_sparse that participates in the device generic test framework. This is important as it will allow us to filter tests by which devices they use, for example, in the future.

Some follow-up tasks:

- I made a few inline comments which should be easy to address, like simplifying the decorator
- It'd be nice if we could switch from using double regularly to using float regularly, since PyTorch's default floating point type is float32

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]

aocsa commented:
Thanks @mruberry for the feedback. I addressed most of the feedback in the new update. 

> Hey @aocsa! Because this is such a large PR I'd like to land it quickly, so I propose we stop editing it now, get it landed, and then we can deal with the few follow-up issues separately. Does that sound good to you?

Yeap, that sounds good to me 

> 
> It's great to see a more modern version of test_sparse that participates in the device generic test framework. This is important as it will allow us to filter tests by which devices they use, for example, in the future.
> 
> Some follow-up tasks:
> 
> * I made a few inline comments which should be easy to address, like simplifying the decorator

I followed your advice and now is simpler 

> * It'd be nice if we could switch from using double regularly to using float regularly, since PyTorch's default floating point type is float32

I was looking this, so it looks like that  the default dtype for `test_sparse`   is `torch.double` for a reason, some tests only work properly when dtype is 'torch.double' and not 'torch.float'. 
https://github.com/pytorch/pytorch/blob/0e03a2978ad5ed2be09f3e4691b2d04d6d4c6c90/test/test_sparse.py#L5

The reason is that the test contains some gradchecks and  for torch.float with default eps/tolerance will fail.

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]

mruberry commented:
This just needs to fix the lint error:

```
test/test_autograd.py:7072: 
The above files have trailing spaces; please remove them
```

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]

aocsa commented:
> This just needs to fix the lint error:
> 
> ```
> test/test_autograd.py:7072: 
> The above files have trailing spaces; please remove them
> ```

it's done, thanks @mruberry

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]

aocsa committed:
Update on "modernize test_sparse"


Adding device generic tests to `test_sparse`. 
Follow-up PR: #54153

I think is ready to review. 
Looking forward your comments cc @mruberry. 

Thanks

Differential Revision: [D27562663](https://our.internmc.facebook.com/intern/diff/D27562663)

[ghstack-poisoned]
